packages/blade/src/components/FileUpload/FileUpload.web.tsx

```diff
@@ -0,0 +1,323 @@
+import { FileUploadItem } from './FileUploadItem';
+import BaseBox from '~components/Box/BaseBox';
+import { Box } from '~components/Box';
+import { SelectorLabel } from '~components/Form/Selector/SelectorLabel';
+import { SelectorInput } from '~components/Form/Selector/SelectorInput';
+import { screenReaderStyles } from '~components/VisuallyHidden';
+import { FormHint, FormLabel } from '~components/Form';
+import { useFormId } from '~components/Form/useFormId';
+import { assignWithoutSideEffects } from '~utils/assignWithoutSideEffects';
+import { metaAttribute, MetaConstants } from '~utils/metaAttribute';
+import { getStyledProps } from '~components/Box/styledProps';
+import { makeSize, useTheme } from '~utils';
+import { Text } from '~components/Typography';
+import type { BladeElementRef } from '~utils/types';
+import { getHintType } from '~components/Input/BaseInput/BaseInput';
+import { makeAccessible } from '~utils/makeAccessible';
+
+const MemoizedFileUploadItem = memo(FileUploadItem);
+
+const _FileUpload: React.ForwardRefRenderFunction<BladeElementRef, FileUploadProps> = (
+  {
+    name,
+    accept,
+    uploadType = 'single',
+    onChange,
+    onPreview,
+    onRemove,
+    onCancel,
+    onDrop,
+    isDisabled,
+    isRequired,
+    necessityIndicator,
+    fileList,
+    testID,
+    label,
+    labelPosition = 'top',
+    accessibilityLabel,
+    validationState,
+    helpText,
+    errorText,
+    maxCount,
+    maxSize,
+    ...styledProps
+  },
+  ref,
+): React.ReactElement => {
+  const { platform } = useTheme();
+  const [selectedFiles, setSelectedFiles] = useState<BladeFileList>(fileList ?? []);
+  const [errorMessage, setErrorMessage] = useState(errorText);
+  const [internalValidationState, setInternalValidationState] = useState('none');
+  const [isActive, setIsActive] = useState(false);
+
+  const isMultiple = uploadType === 'multiple';
+  const inputLabelPosition = platform === 'onMobile' ? 'top' : labelPosition;
+  const isLabelLeftPositioned = inputLabelPosition === 'left';
+  const willRenderHintText =
+    !isMultiple && selectedFiles.length === 1 ? false : Boolean(helpText) || Boolean(errorMessage);
+
+  const showError = validationState === 'error' || internalValidationState === 'error';
+  const showHelpText = !showError && helpText;
+  const accessibilityText =
+    accessibilityLabel ?? `,${showError ? errorText : ''} ${showHelpText ? helpText : ''}`;
+  const { inputId, labelId, helpTextId, errorTextId } = useFormId('fileinput');
+
+  const accessibilityProps = makeAccessible({
+    required: Boolean(isRequired),
+    invalid: Boolean(showError),
+    disabled: Boolean(isDisabled),
+    describedBy: helpTextId,
+  });
+
+  // In control mode attach a unique id to each file if not provided
+  useMemo(() => {
+    for (const file of selectedFiles) {
+      if (!file.id) {
+        file.id = `${new Date().getTime().toString()}${Math.floor(Math.random() * 1000000)}`;
```

Comment:
We are mutating the files without setState?

Reply:
If we use setState it will go in an infinite re-render loop.

Reply:
No, hooks can't be called inside a callback.